### ISSUE TITLE
Error updating metadata job

### ISSUE THREAD

**Issue by camillo-positano** [JMBeresford/retrom#135]
I have retrom on my server behind portainer and nginx as a reverse proxy. 

I successfully connected to the server and setup both server and client container. 
When i execute the "Download Metadata" from the ui nothing happens and in the container of the BE i get the following logs:

```
2024-09-27T22:25:51.290106Z  INFO update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Platform Metadata" opts=None}:job_thread: retrom_service::grpc::jobs::job_manager: Job completed: Downloading Platform Metadata

2024-09-27T22:25:51.290216Z  INFO update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Game Metadata" opts=Some(JobOptions { wait_on_jobs: Some([c7334a14-82a0-43c6-9341-e69a0ccb358f]) })}:job_thread: retrom_service::grpc::jobs::job_manager: Job completed: Downloading Game Metadata

2024-09-27T22:25:51.290289Z  INFO update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Extra Metadata" opts=Some(JobOptions { wait_on_jobs: Some([323ffa75-3878-4115-814b-ba74a55dcddf]) })}:job_thread: retrom_service::grpc::jobs::job_manager: Job completed: Downloading Extra Metadata

2024-09-27T22:25:51.290337Z ERROR update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Extra Metadata" opts=Some(JobOptions { wait_on_jobs: Some([323ffa75-3878-4115-814b-ba74a55dcddf]) })}:job_thread: retrom_service::grpc::jobs::job_manager: Invalidation channel closed: SendError(6e7df925-acf9-4fed-a638-01817df0d646)
```

Some info:

- I have the database on a separate container behind its own sub domain with a dedicated user and password.

- This is my config file:

```
{
  "connection": {
    "port": 5101,
    "db_url": "postgres://***:***]@***"
  },
  "content_directories": [
    {
      "path": "/library1",
      "storage_type": "MultiFileGame"
    }
  ],
  "igdb": {
    "client_secret": "***",
    "client_id": "***"
  }
}
```

(I redacted what is considered sensitive from the config file)

- Screenshot of portainer:
![immagine](https://github.com/user-attachments/assets/078b6d06-709f-4bf9-9dbb-7dc892d5aa5f)

Am i missing something?

**Comment by camillo-positano** [comment]
Also i have the database on a separate container behind its own sub domain with a dedicated user and password

**Comment by JMBeresford** [comment]
I assume your library populated successfully, and it is only the metadata job that is doing nothing as you did not mention that? Note that you need to run the Update Library job first.

Can you add `RUST_LOG=debug` to the env for the container, restart it and try again w/ debug logs

**Comment by camillo-positano** [comment]
I did indeed try the update first but i got a popup saying: 
![immagine](https://github.com/user-attachments/assets/fc99f7eb-ce84-42a9-b915-3f9ab17796a3)

So i assumed that it's because i have nothing in it, and that i had to retrieve metadata to download content into it.

Am i thinking wrong?

Thank you for the quick response.
I'll try and increase the logging and report back to you with more info

**Comment by camillo-positano** [comment]
This is the additional logs i got:
```
2024-09-27T23:02:56.849689Z  INFO retrom_service: Starting server at: 0.0.0.0:5101

2024-09-27T23:03:29.785628Z DEBUG update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Platform Metadata" opts=None}: retrom_service::grpc::jobs::job_manager: No recievers for new job: "Downloading Platform Metadata"

2024-09-27T23:03:29.785650Z DEBUG update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Game Metadata" opts=Some(JobOptions { wait_on_jobs: Some([dd996012-2edf-4e92-8295-fd34587627fc]) })}: retrom_service::grpc::jobs::job_manager: No recievers for new job: "Downloading Game Metadata"

2024-09-27T23:03:29.785727Z  INFO update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Platform Metadata" opts=None}:job_thread: retrom_service::grpc::jobs::job_manager: Job completed: Downloading Platform Metadata

2024-09-27T23:03:29.785747Z  INFO update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Game Metadata" opts=Some(JobOptions { wait_on_jobs: Some([dd996012-2edf-4e92-8295-fd34587627fc]) })}:job_thread: retrom_service::grpc::jobs::job_manager: Job completed: Downloading Game Metadata

2024-09-27T23:03:29.785761Z  INFO update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Extra Metadata" opts=Some(JobOptions { wait_on_jobs: Some([05d2f00a-ee1e-4663-b203-d244609db774]) })}:job_thread: retrom_service::grpc::jobs::job_manager: Job completed: Downloading Extra Metadata

2024-09-27T23:03:29.785767Z ERROR update_library_metadata:update_metadata{overwrite=false}:spawn{name="Downloading Extra Metadata" opts=Some(JobOptions { wait_on_jobs: Some([05d2f00a-ee1e-4663-b203-d244609db774]) })}:job_thread: retrom_service::grpc::jobs::job_manager: Invalidation channel closed: SendError(80624dad-70a0-4c53-9252-079d7ae10a48)

2024-09-27T23:03:29.801296Z DEBUG get_jobs:get_jobs_stream: retrom_service::grpc::jobs: Jobs stream ended

2024-09-27T23:03:29.801307Z DEBUG get_job_subscription:get_jobs_stream: retrom_service::grpc::jobs: Jobs stream ended

2024-09-27T23:03:29.820676Z DEBUG get_job_subscription:get_jobs_stream: retrom_service::grpc::jobs: Jobs stream ended

2024-09-27T23:03:29.850609Z DEBUG get_job_subscription:get_jobs_stream: retrom_service::grpc::jobs: Jobs stream ended
```

**Comment by JMBeresford** [comment]
Ah, it cannot find the library dir. Does the container have at least read permissions for the `retrom_library` volume? By default the container runs as `1001:1001` (changing to `1505:1505` in newest update), but you can run it w/ any uid:gid combo you want that match the permissions you need to access the host files.

i.e. if on the host machine only `1000:1000` has read access to your library, then the same applies in the container -- so you need to either run the container w/ valid IDs or you need to update the permissions for the files on the host.

It could also be that your config file is not being loaded for some reason, and thus the content_dir is not being looked for. Can you exec into the container and check both for `/library1` and `/config` and their contents?

**Comment by camillo-positano** [comment]
i did indeed check the config file and it matches, also the library directory is inside the container.

I tried creating a file into the library path inside the container and i managed to successfully create it and it reflected on my host machine, so i dont this its a volume permission

**Comment by camillo-positano** [comment]
Also copy pasted the path from the config inside the container to ls and create/delete a file and it had no problem, so i would imagine the config file is wel configured

**Comment by JMBeresford** [comment]
And the config file path is _exactly_ `/config/config.json`? What debug logs are present when you run the `Update Library` job?

**Comment by camillo-positano** [comment]
Yeah this it the ls from the container:

![immagine](https://github.com/user-attachments/assets/c68d39fa-534e-4b82-af87-710e88aad791)

There are no logs when i execute the Update Library request

**Comment by camillo-positano** [comment]
The network of the client executes a get and a response 200 with no body is returned

**Comment by JMBeresford** [comment]
That's to be expected w/ an empty content_dir config. You have verified the library directory is in the correct structure? e.g. (`/library1/game_boy/pokemon_red/pokemon_red.gba`)

What happens if you remove the `RETROM_CONFIG` env and instead populate:

- IGDB_CLIENT_ID
- IGDB_CLIENT_SECRET
- DB_URL

With the relevant entries. Legacy behavior is for `/app/library` to act as the default content_directory -- so remap your volume there when you test this.

**Comment by camillo-positano** [comment]
The library content is and empty directory, is that not correct? i read the directory structure, but i thought that was for when you had already a game

**Comment by camillo-positano** [comment]
I'll try the "legacy" tomorrow, because where i am now it's 1.30 AM.

Thanks for the support!

**Comment by JMBeresford** [comment]
Ah, well yes you need games in there for any of these jobs to find them. I do however think the error should indicate the directory is empty, rather than unfound.

Add your roms to the directory as described and things should work just find. May I ask what metadata you were expecting to download w/ an empty library? I am curious if my docs are perhaps misleading.

**Comment by JMBeresford** [comment]
Note that Retrom simply takes your existing library and presents it for use in the various clients. It does not facilitate any sort of acquisition of games/roms.

**Comment by camillo-positano** [comment]
Aha! Then i completely misunderstood the purpose of the repo sorry! I thought that by metadata it also included a remote url from which to download the rom from and save it into your library.

My bad! 

Thank you very much for the help tho!

**Comment by JMBeresford** [comment]
No worries, I hope you can still find use for Retrom someday!

**Comment by Szymok** [comment]
I have similar error, but i insert some games into the library to test it. 

szymon@firebat:~/retrom/library/gba$ ls
'Pokemon Pinball - Ruby & Sapphire (Europe) (En,Fr,De,Es,It).gba'

BUt during refreshing the library in docker container logs i received error:
`2024-09-30T18:19:56.444612930Z 2024-09-30T18:19:56.444488Z  WARN update_library:update_library: retrom_service::grpc::library::update_handlers: Failed to resolve platforms: Could not read directory

2024-09-30T18:19:56.447684986Z 2024-09-30T18:19:56.447497Z  INFO update_library:update_library:spawn{name="Update Library" opts=None}:job_thread: retrom_service::grpc::jobs::job_manager: Job completed: Update Library

2024-09-30T18:19:56.447727739Z 2024-09-30T18:19:56.447533Z ERROR update_library:update_library:spawn{name="Update Library" opts=None}:job_thread: retrom_service::grpc::jobs::job_manager: Invalidation channel closed: SendError(f1b23bc6-641f-45a5-bd10-545fe35624df)`

I was trying to find any mistakes with my permissions but i have inside the docker permission and outside to the folders, so i do not have any idea right now.

**Comment by JMBeresford** [comment]
@Szymok could you open a new issue for this, please include the following:

- Server config file
- Docker compose file (**with secrets redacted**)
- Evidence/screenshots of your library being available in the container (e.g. exec into the container and inspect the directory you mounted it to and make sure the files are there)
